include/quic/endpoint.hpp

```diff
@@ -17,56 +17,118 @@ extern "C"
+        std::shared_ptr<uv_udp_t> handle;
+        bool accepting_inbound{false};
+        Network& net;
 
-        std::shared_ptr<Handler> handler;
+      public:
+        explicit Endpoint(Network& n, const Address& listen_addr);
+
+        template <typename... Opt>
+        bool listen(Opt&&... opts)
+        {
+            std::promise<bool> p;
+            auto f = p.get_future();
+
+            net.call([&opts..., &p, this]() mutable {
+                try
+                {
+                    // initialize client context and client tls context simultaneously
+                    inbound_ctx = std::make_shared<InboundContext>(std::forward<Opt>(opts)...);
+                    accepting_inbound = true;
+
+                    log::debug(log_cat, "Inbound context ready for incoming connections");
+
+                    p.set_value(true);
+                }
+                catch (...)
+                {
+                    p.set_exception(std::current_exception());
+                }
+            });
+
+            return f.get();
+        }
+
+        // creates new outbound connection to remote; emplaces conn/interface pair in outbound map
+        template <typename... Opt>
+        std::shared_ptr<connection_interface> connect(const Address& remote, Opt&&... opts)
+        {
+            std::promise<std::shared_ptr<Connection>> p;
+            auto f = p.get_future();
+
+            net.call([&opts..., &p, this, raddr = remote]() mutable {
+                try
+                {
+                    // initialize client context and client tls context simultaneously
+                    outbound_ctx = std::make_shared<OutboundContext>(std::forward<Opt>(opts)...);
+
+                    if (auto [itr, res] = conns.emplace(ConnectionID::random(), nullptr); res)
+                    {
+                        itr->second = std::move(Connection::make_conn(
+                                *this,
+                                itr->first,
+                                ConnectionID::random(),
+                                Path{local, raddr},
+                                handle,
+                                outbound_ctx->tls_creds,
+                                outbound_ctx->config,
+                                Direction::OUTBOUND));
+
+                        p.set_value(itr->second);
+                    }
+                    else
+                        p.set_value(nullptr);
+                }
+                catch (...)
+                {
+                    p.set_exception(std::current_exception());
+                }
+            });
+
+            return f.get();
+        };
 
         std::shared_ptr<uvw::loop> get_loop();
 
+        // query a list of all active inbound and outbound connections paired with a conn_interface
+        std::list<std::pair<ConnectionID, std::shared_ptr<connection_interface>>> get_all_conns(std::optional<Direction> d = std::nullopt);
```

Comment:
I think this can just return a `list<shared_ptr<connection_interface>>`: you should be able to retrieve the ConnectionID out of the connection_interface.